pkg/multicast/multicast.go

```diff
@@ -411,149 +420,139 @@ func (c *Client) Handle(m *sbe.SbeGoMarshaller, r io.Reader) error {
 		return err
 	}
 
-	for _, event := range events {
-		switch event.Type {
-		case EventTypeInstrument:
-			// update instrumentsMap
-			ins, ok := event.Data.(models.Instrument)
-			if !ok {
-				return fmt.Errorf("invalid event type")
-			}
-			c.setInstrument(ins.InstrumentID, ins)
-
-			// emit event
-			c.Emit(newInstrumentNotificationChannel(ins.Kind, ins.BaseCurrency), &event.Data)
-			c.Emit(newInstrumentNotificationChannel(KindAny, ins.BaseCurrency), &event.Data)
+	c.emitEvents(events)
+	return nil
+}
 
-		case EventTypeOrderBook:
-			books := event.Data.(models.OrderBookRawNotification)
-			c.Emit(newOrderBookNotificationChannel(books.InstrumentName), &event.Data)
+type Bytes []byte
+type Pool struct {
+	mu             *sync.RWMutex
+	queue          []Bytes
+	maxPackageSize int
+}
 
-		case EventTypeTrades:
-			trades := event.Data.(models.TradesNotification)
-			if len(trades) > 0 {
-				tradeIns := trades[0].InstrumentName
-				tradeKind := trades[0].InstrumentKind
-				currency := getCurrencyFromInstrument(tradeIns)
-				c.Emit(newTradesNotificationChannel(tradeKind, currency), &event.Data)
-			}
+func NewPool(maxPackageSize int) *Pool {
+	q := make([]Bytes, 0)
 
-		case EventTypeTicker:
-			ticker := event.Data.(models.TickerNotification)
-			c.Emit(newTickerNotificationChannel(ticker.InstrumentName), &event.Data)
-		}
+	return &Pool{
+		mu:             &sync.RWMutex{},
+		queue:          q,
+		maxPackageSize: maxPackageSize,
 	}
-
-	return nil
 }
 
-func (c *Client) getConnection(addr string) *net.UDPConn {
-	c.mu.Lock()
-	defer c.mu.Unlock()
+func (p *Pool) Get() Bytes {
+	p.mu.Lock()
+	defer p.mu.Unlock()
 
-	return c.addrConnMap[addr]
+	if length := len(p.queue); length > 0 {
+		item := p.queue[length-1]
+		p.queue[length-1] = nil
+		p.queue = p.queue[:length-1]
+		return item
+	}
+	return make(Bytes, p.maxPackageSize)
 }
 
-func (c *Client) setConnection(addr string, conn *net.UDPConn) {
-	c.mu.Lock()
-	defer c.mu.Unlock()
+func (p *Pool) Put(bs Bytes) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
 
-	c.addrConnMap[addr] = conn
+	p.queue = append(p.queue, bs)
 }
 
-func (c *Client) delConnection(addr string) {
-	c.mu.Lock()
-	defer c.mu.Unlock()
-
-	delete(c.addrConnMap, addr)
-}
+// ListenToEvents listens to a list of udp addresses on given network interface.
+func (c *Client) ListenToEvents(ctx context.Context) error {
+	packetConn, err := net.ListenPacket("upd4", fmt.Sprintf("0.0.0.0:%d", c.port))
 
-func (c *Client) listenToMulticastUDP(addr string) (*net.UDPConn, error) {
-	udpAddr, err := net.ResolveUDPAddr("udp", addr)
 	if err != nil {
-		c.log.Errorw("failed to resolve UDP address", "err", err)
-		return nil, err
+		c.log.Errorw("failed to initiate packet connection", "err", err, "port", c.port)
+		return err
 	}
+	defer packetConn.Close()
```

Comment:
Should not `defer packetConn.Close()` here, because we also set the connection to structure field.